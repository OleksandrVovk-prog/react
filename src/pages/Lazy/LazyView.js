import { lazy, Suspense } from 'react';
import PropTypes from 'prop-types';
import Button from '../../components/Button/Button';
import styles from './sass/Lazy.module.scss';

const Film = lazy(() => import('../../components/Film/Film'));

export default function LazyView({
  information, getFilm, title, buttonTitle, loading, search, handleSearch
}) {
  return (
    <div>
      <p className={styles.appIntro}>{title}</p>
      <Suspense fallback={<p>Loading component...</p>}>
        {
          loading
            ? <p>Loading...</p>
            : information.Title && <Film film={information} />
        }
      </Suspense>
      <input type="text" value={search} onChange={handleSearch} data-testid="input-search" />
      <Button
        title={buttonTitle}
        buttonClass={styles.updateFilm}
        handleClick={getFilm}
        id="update-film-button"
      />
    </div>
  );
}

LazyView.propTypes = {
  information: PropTypes.shape({
    Title: PropTypes.string
  }),
  getFilm: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  buttonTitle: PropTypes.string.isRequired,
  loading: PropTypes.bool.isRequired,
  search: PropTypes.string.isRequired,
  handleSearch: PropTypes.func.isRequired,
};

LazyView.defaultProps = {
  information: {}
};
