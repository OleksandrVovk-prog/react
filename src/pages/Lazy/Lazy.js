import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import LazyView from './LazyView';

function Lazy() {
  const { t } = useTranslation();
  const [information, setInformation] = useState({});
  const [search, setSearch] = useState('');
  const [loading, setLoading] = useState(false);

  const getFilm = () => {
    setLoading(true);
    fetch(`https://www.omdbapi.com/?apikey=3db238d1&t=${search}`).then((res) => {
      if (res.ok) {
        res.json().then((film) => {
          setInformation(film);
          setLoading(false);
        });
      }
    }).catch(() => setLoading(false));
  };

  const handleSearch = (event) => {
    setSearch(event.target.value);
  };

  return (
    <LazyView
      information={information}
      getFilm={getFilm}
      title={t('lazyPage')}
      buttonTitle={t('search')}
      loading={loading}
      search={search}
      handleSearch={handleSearch}
    />
  );
}

export default Lazy;
