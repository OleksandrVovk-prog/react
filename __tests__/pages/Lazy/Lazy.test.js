import React from 'react';
import userEvent from '@testing-library/user-event';
import fetchMock from 'jest-fetch-mock';
import { render } from '../../settings/test-utils';
import Lazy from '../../../src/pages/Lazy/Lazy';

describe('<Lazy />', () => {
  let wrapper;
  const mockDataResponse = {
    Title: 'mock title'
  };
  const mockDataReject = 'mock data reject';
  beforeEach(() => {
    fetchMock.mockResponseOnce(JSON.stringify(mockDataResponse));
    fetchMock.mockRejectOnce(JSON.stringify(mockDataReject));
    wrapper = render(<Lazy />);
  });

  test('should render Lazy component', () => {
    const { asFragment } = wrapper;
    expect(asFragment(wrapper)).toMatchSnapshot();
  });

  test('should change input value', async () => {
    const { getByTestId } = wrapper;
    const Input = getByTestId(/input-search/i);
    await userEvent.type(Input, 'mock value');
    expect(Input.value).toEqual('mock value');
  });
  test('should change loading value to true', async () => {
    const { getByTestId, getByText } = wrapper;
    const Button = getByTestId('update-film-button');
    await userEvent.click(Button);
    expect(getByText(/Loading.../i)).toBeInTheDocument();
  });
});
