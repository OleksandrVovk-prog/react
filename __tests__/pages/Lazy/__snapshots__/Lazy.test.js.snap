// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Lazy /> should render Lazy component 1`] = `
<DocumentFragment>
  <div>
    <p
      class="appIntro"
    >
      Lazy page
    </p>
    <input
      data-testid="input-search"
      type="text"
      value=""
    />
    <button
      class="updateFilm"
      data-testid="update-film-button"
      type="button"
    >
      Search
    </button>
  </div>
</DocumentFragment>
`;
